src/v/cluster/archival/archival_policy.cc

```diff
@@ -185,18 +185,6 @@ archival_policy::lookup_result archival_policy::find_segment(
     const auto& set = log->segments();
     const auto& ntp_conf = log->config();
     auto it = set.lower_bound(start_offset);
-    if (it == set.end() || eligible_for_compacted_reupload(**it)) {
-        // Skip forward if we hit a gap or compacted segment
-        for (auto i = set.begin(); i != set.end(); i++) {
-            const auto& sg = *i;
-            if (start_offset < sg->offsets().get_base_offset()) {
-                // Move last offset forward
-                it = i;
-                start_offset = sg->offsets().get_base_offset();
-                break;
-            }
-        }
-    }
```

Comment:
This clause used to take care of these conditions:

* if the `start_offset` lies in a gap, then make `it` point to the segment after the gap
* if `it` is compacted, then point to the segment starting after `start_offset` (which might also be compacted)

From the cover letter it is not possible for this code path to encounter a compacted segment because of the condition imposed on compaction (no compaction before upload). This path is also not  used for compacted re-upload. 

That still leaves the case where the `start_offset` lies in a gap. Do we no longer expect this to happen? From what I can see now the progress will stop if the offset lies in a gap between two segments.

Reply:
If the `start_offset` lies in the gap the `lower_bound` will find the next segment in the list (which will have base-offset >= start_offset).

Reply:
It looked to me (the code is a little convoluted in segment set) like it will return `end` because from https://github.com/redpanda-data/redpanda/blob/dev/src/v/storage/segment_set.cc#L94-L113 both `needle_in_range` checks after the segment after the offset is selected, will fail because the offset is in neither the segment right after the gap nor in the segment before the gap. 

But if it's known that the next segment will be returned and not `end` then that sounds fine.

Reply:
I added the alternative implementation for gap handling. But the code now doesn't care if the segment compacted or not.